autotyping/autotyping.py

```diff
@@ -535,40 +549,57 @@ def leave_Param(
                     )
                 )
 
+        parameter_name = original_node.name.value
         default_is_none = (
             original_node.default is not None
             and isinstance(original_node.default, libcst.Name)
             and original_node.default.value == "None"
         )
+        # default value is None, i.e. `def foo(bar=None)`
         if default_is_none:
+            # check if user has explicitly specified a type for this name
             for anno_optional in self.state.annotate_optionals:
-                if original_node.name.value == anno_optional.name:
+                if parameter_name == anno_optional.name:
                     return self._annotate_param(
-                        anno_optional, updated_node, optional=True
+                        anno_optional, updated_node, containers=["Optional"]
                     )
+            # guess type from name
+            if self.state.guess_common_names:
+                guessed_param, containers = guess_type_from_argname(parameter_name)
+                if guessed_param:
+                    return self._annotate_param(
+                        guessed_param, updated_node, containers + ["Optional"]
+                    )
+
+        # no default value, i.e. `def foo(bar)`
         elif original_node.default is None:
+            # check if user has explicitly specified a type for this name
             for anno_named_param in self.state.annotate_named_params:
                 if original_node.name.value == anno_named_param.name:
-                    return self._annotate_param(anno_named_param, updated_node)
+                    return self._annotate_param(anno_named_param, updated_node, [])
+            # guess type from name
+            if self.state.guess_common_names:
+                guessed_param, containers = guess_type_from_argname(parameter_name)
+                if guessed_param:
+                    return self._annotate_param(guessed_param, updated_node, containers)
+
         return updated_node
 
     def _annotate_param(
-        self, param: NamedParam, updated_node: libcst.Param, optional: bool = False
+        self, param: NamedParam, updated_node: libcst.Param, containers: List[str]
     ) -> libcst.Param:
         if param.module is not None:
             AddImportsVisitor.add_needed_import(
                 self.context, param.module, param.type_name
             )
-        if optional:
-            AddImportsVisitor.add_needed_import(self.context, "typing", "Optional")
-        type_name = libcst.Name(value=param.type_name)
-        if optional:
+        anno = libcst.Name(value=param.type_name)
+
+        for container in containers:
+            AddImportsVisitor.add_needed_import(self.context, "typing", container)
```

Comment:
Assuming all the containers are from `typing` works for now but isn't correct in general. For example, once we're 3.9+ we should generate `list[int]`. (And maybe we should already have an option to do that for codebases that are already 3.9+.) However, this would add some complexity. I think for now let's stick with what you have, then add a set of options like `--py39`, `--py310` to affect autotyping's output.

Reply:
I'd be inclined to just recommend `pyupgrade` to your users for that, rather than building it into `autotyping` at this stage, and then doing a hard cutover when 3.8 reaches EOL.  But ofc your call!

Reply:
I also thought about this and generally agree with recommending pyupgrade. Though given that the code for it should be written for it at some point or another it doesn't really hurt (other than some code complexity) to write that now and gate the functionality behind a command line parameter.